crates/snforge-scarb-plugin/src/attributes/available_gas.rs

```
@@ -24,14 +25,32 @@ impl AttributeCollector for AvailableGasCollector {
         args: Arguments,
         _warns: &mut Vec<Diagnostic>,
     ) -> Result<String, Diagnostics> {
-        let &[arg] = args.unnamed_only::<Self>()?.of_length::<1, Self>()?;
-
-        let gas = Number::parse_from_expr::<Self>(db, arg.1, arg.0.to_string().as_str())?;
-
-        let gas = gas.as_cairo_expression();
+        let named_args = args.named_only::<Self>()?;
+
+        let l1_gas = named_args
+            .as_once_optional("l1_gas")?
+            .map(|arg| Number::parse_from_expr::<Self>(db, arg, "l1_gas"))
+            .transpose()?
+            .unwrap_or(Number(BigInt::ZERO));
```

Shouldn't the logic rather be: if not provided treat it as unlimited? I expect this to be a common use case to limit only a specific gas type. With the current approach it also enforces setting other types if they are not 0.

---

Ahh yes, you're right that is just plain better UX. 
the only caveat is, right now technically you can use l1 gas to pay for execution resources (normally paid for by l2), so you will need to be very careful with setting those two (eg if you were to set l2 to something that is not sufficient, l1 can be taken and chipped in to cover the fee). I believe this behaviour will not be permitted from next starknet version.

This can be pretty confusing, I think. Thoughts?

---

Maybe if we could allow explicitly setting these to unlimited somehow (maybe through some kind of enum, but not sure if this is possible in plugin), then it would be the best. 
But that might be to complicated to implement.

---

> I believe this behaviour will not be permitted from next starknet version

Do you mean from `0.14.0`? Can we quickly verify if setting `l2_gas` to a very low value with logic for unlimited `l1_gas` will result in the missing gas being taken from l1?

> Maybe if we could allow explicitly setting these to unlimited somehow 

The default behavior is unlimited, so for consistency not setting a given type should keep it as the default, so unlimited.

---

Yes, I meant 0.14. To verify this, I have modified `integration::available_gas::available_gas_exceeded` test from `forge` crate; currently it has only l2_gas set to 5, the rest is set to 0 (default), and the test checks if a panic is being thrown because available gas was exceeded (it does). Now I change `available_gas` like so:

FROM:
```
            #[test]
            #[available_gas(l2_gas: 5)]
            fn keccak_cost() {
                keccak::keccak_u256s_le_inputs(array![1].span());
            }
```

TO:
```
            #[test]
            #[available_gas(l1_gas: 999999999999999999999999999999999, l2_gas: 5)]
            fn keccak_cost() {
                keccak::keccak_u256s_le_inputs(array![1].span());
            }
```

and the tests fails with:

```
thread 'integration::available_gas::available_gas_exceeded' panicked at crates/forge/test_utils/src/runner.rs:253:5:
Some tests didn't fail
```

upon inspection, to make sure it fails because test succeeds and not something totally unrelated, I see:

```
    TestTargetSummary {
        test_case_summaries: [
            Single(
                Passed {
                    name: "test_package_integrationtest::test_case::keccak_cost",
```

> The default behavior is unlimited, so for consistency not setting a given type should keep it as the default, so unlimited.

This was far easier, when only max_fee could be specified 😞 
I mean, I do not see a clear solution for this to be honest (apart from setting this to 0 as it is right now or just straight up requiring users to set all bounds)

---

The way we handle max fee is to verify after the test execution is completed if the value didn't exceed max fee. Couldn't we type these bounds as `Option` inside forge and treat them as unlimited when checking gas if they are `None`?

---

> upon inspection, to make sure it fails because test succeeds and not something totally unrelated

Tbh it doesn't make much sense to me looking at the current logic [here](https://github.com/foundry-rs/starknet-foundry/pull/3054/files#diff-5f4c6e6bd283c1d085b4ec4d8dc055fcdb9f53820fa682274675aef306f3be7dR163) and actually the example above doesn't really check if the available gas was exceeded, it only checks whether this test fails or not. My guess it that the number you provided above for `l1_gas` exceeds `u64` and therefore it silently fails conversion to `GasVector` without asserting available gas at the end, but I am not sure. Could you check it again? 🙏

---

lol, you might be onto something with the overflow, great catch 🤔 that means there is a bug somewhere in the code (my guess is attributes?) as **every test just passes if you provide a number big enough**, for any field - that is wild (I just confirmed it with latest released snforge as well, but since it used max_fee it probably was not that big of a problem up until now). Will try to find it, and keep you updated.

The question with paying with l1 for l2 still remains (and will result in the behaviour I described) - I think it's time for a decision - do we want it or not? connects to this thread https://github.com/foundry-rs/starknet-foundry/pull/3054#discussion_r1987593228

---

Ok, so I found the issue - Once again thanks Darek for diggint into this, I wouldn't have found it otherwise 🙏 

The plugin took a number from the attribute as felt, and then it was casted to BigInt. Then, forge took it and passed it to the vm, which failed silently (tldr it couldn't fit bigger values into GasAmount, so it defaulted to None and returned it to the forge). Forge then operated on a config, where available gas was not set (or rather - been set to None, instead of the value from the attribute), that's why the tests always passed.

This bug was there from the beggining, I guess no-one ever spotted this because up until now, if someone set max_fee larger that u64::MAX, it might as well be infinite. The problem manifested itself when using bounds, as setting one of the bounds as larger than u64::MAX nullified all other bounds, making the gas infinite for all of them.

This is now fixed on the plugin side, values are verified for both max_fee and bounds cases (although you can probably make a case to not doing it for max_fee, as the issue is not that severe there).

Regarding the second part of it, I have made sure for now that no conversion of l1 <-> l2 for payment will be made implicitly, so we should be good for time being and decide what to do next when starknet 0.14 is up.